src/Persistence/Sql.php

```diff
@@ -364,60 +364,36 @@ protected function setLimitOrder(Model $m, Query $q)
+     * Will apply scope defined inside $scope or $model->scope() onto $query.
      */
-    public function initQueryConditions(Model $m, Query $q): Query
+    public function initQueryConditions(Model $model, Query $query, Model\Scope\AbstractCondition $condition = null): Query
     {
-        if (!isset($m->conditions)) {
-            // no conditions are set in the model
-            return $q;
-        }
-
-        foreach ($m->conditions as $cond) {
-            // Options here are:
-            // count($cond) == 1, we will pass the only
-            // parameter inside where()
-
-            if (count($cond) === 1) {
-                // OR conditions
-                if (is_array($cond[0])) {
-                    foreach ($cond[0] as &$row) {
-                        if (is_string($row[0])) {
-                            $row[0] = $m->getField($row[0]);
-                        }
-
-                        // "like" or "regexp" conditions do not need typecasting to field type!
-                        if ($row[0] instanceof Field && (count($row) === 2 || !in_array(strtolower($row[1]), ['like', 'regexp'], true))) {
-                            $valueKey = count($row) === 2 ? 1 : 2;
-                            $row[$valueKey] = $this->typecastSaveField($row[0], $row[$valueKey]);
-                        }
-                    }
-                }
+        $condition = $condition ?? $model->scope();
 
-                $q->where($cond[0]);
+        if (!$condition || $condition->isEmpty()) {
+            return $query;
+        }
 
-                continue;
-            }
+        // peel off the single nested scopes to convert (((field = value))) to field = value
+        $condition = $condition->simplify();
 
-            if (is_string($cond[0])) {
-                $cond[0] = $m->getField($cond[0]);
-            }
+        // simple condition
+        if ($condition instanceof Model\Scope\BasicCondition) {
+            $query = $query->where(...$condition->toArray());
+        }
 
-            if (count($cond) === 2) {
-                if ($cond[0] instanceof Field) {
-                    $cond[1] = $this->typecastSaveField($cond[0], $cond[1]);
-                }
-                $q->where($cond[0], $cond[1]);
-            } else {
-                // "like" or "regexp" conditions do not need typecasting to field type!
-                if ($cond[0] instanceof Field && !in_array(strtolower($cond[1]), ['like', 'regexp'], true)) {
-                    $cond[2] = $this->typecastSaveField($cond[0], $cond[2]);
-                }
-                $q->where($cond[0], $cond[1], $cond[2]);
+        // nested conditions
+        if ($condition instanceof Model\Scope\CompoundCondition) {
+            $expression = $condition->isOr() ? $query->orExpr() : $query->andExpr();
+
+            foreach ($condition->getNestedConditions() as $nestedCondition) {
+                $expression = $this->initQueryConditions($model, $expression, $nestedCondition);
```

Comment:
some usages of `initQueryConditions()` does not assign, I propose to change `initQueryConditions()` to return void

Reply:
done